### PR TITLE
Persistent queue

### PR DESCRIPTION
**Pull request by wojtekzyla**
Things included in this PR:

1. enable persistent queues configuration in the helm chart

Continuation of this PR updated with changes from current main branch: https://github.com/signalfx/splunk-otel-collector-chart/pull/675

### CODE REVIEW

**Comment by jvoravong** [comment]
Can you add a changelog entry?

File: helm-charts/splunk-otel-collector/templates/config/_common.tpl [hunk 1]
```diff
@@ -288,6 +280,16 @@ splunk_hec/platform_traces:
     queue_size: {{ .Values.splunkPlatform.sendingQueue.queueSize }}
 {{- end }}
 
+{{/*
+Splunk sending queue
+*/}}
+{{- define "splunk-otel-collector.splunkSendingQueue" -}}
```
Comment (jvoravong) [inline helm-charts/splunk-otel-collector/templates/config/_common.tpl:286]: If persistent queues were added to splunkObservability components, would they be able to use this? If not, we should rename this here and else where.
splunk-otel-collector.splunkSendingQueue -> splunk-otel-collector.splunkPlatformSendingQueue

File: .gitignore [hunk 2]
```diff
@@ -4,3 +4,4 @@
 # Helm
 **/charts/*.tgz
 helm-charts/splunk-otel-collector/Chart.lock
+.DS_Store
```
Comment (atoulme) [inline .gitignore:7]: nit: make this a separate PR? I ended up myself doing this:
```
> cat ~/.gitignore_global
.idea
.DS_Store
```

File: helm-charts/splunk-otel-collector/templates/config/_common.tpl (hunk 1 above)
Comment (wojtekzyla) [reply helm-charts/splunk-otel-collector/templates/config/_common.tpl]: I renamed this one and other occurrences

Comment (wojtekzyla) [reply helm-charts/splunk-otel-collector/templates/config/_common.tpl]: Also I added a changelog entry

File: .gitignore (hunk 2 above)
Comment (wojtekzyla) [reply .gitignore]: I deleted .DS_Store entry from .gitignore

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 3]
```diff
@@ -639,9 +644,17 @@ exporters:
   {{- end }}
   {{- if (eq (include "splunk-otel-collector.platformLogsEnabled" .) "true") }}
   {{- include "splunk-otel-collector.splunkPlatformLogsExporter" . | nindent 2 }}
+  {{- include "splunk-otel-collector.splunkPlatformSendingQueue" . | nindent 4 }}
+      {{- if .Values.splunkPlatform.sendingQueue.persistentQueueEnabled.logs }}
+      storage: file_storage/persistent_queue
+      {{- end }}
```
Comment (jvoravong) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:650]:
```suggestion
  {{- end }}
```

File: helm-charts/splunk-otel-collector/templates/daemonset.yaml [hunk 4]
```diff
@@ -160,6 +160,11 @@ spec:
           then
               setfacl -n -Rm d:m::rx,m::rx,d:g:{{ $agent.securityContext.runAsGroup | default 999 }}:rx,g:{{ $agent.securityContext.runAsGroup | default 999 }}:rx {{ .Values.logsCollection.journald.directory }};
           fi;
+          {{- end }}
```
Comment (jvoravong) [inline helm-charts/splunk-otel-collector/templates/daemonset.yaml:163]: All the changes in this file will have to have similar changes in deployment-gateway.yaml and deployment-cluster-receiver.yaml

File: helm-charts/splunk-otel-collector/values.yaml [hunk 5]
```diff
@@ -106,6 +106,18 @@ splunkPlatform:
     #   requests_per_second is the average number of requests per seconds.
     queueSize: 5000
 
+    # When enabled, it uses file_storage extension to persist the queue data.
+    # This can be used to prevent data-loss during restart
+    # NOTE: The File Storage extension will persist state to the node's local file system.
+    # While using the persistent queue it is advised to increase memory limit for agent (agent.resources.limits.memory)
+    # to 1Gi.
+    persistentQueueEnabled:
```
Comment (jvoravong) [inline helm-charts/splunk-otel-collector/values.yaml:114]: - We should move these configurations under sendingQueue since they are not different queues.
- We probably need to note the [alpha](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/README.md#persistent-queue) status of this component.

File: helm-charts/splunk-otel-collector/values.yaml [hunk 6]
```diff
@@ -106,6 +106,18 @@ splunkPlatform:
     #   requests_per_second is the average number of requests per seconds.
     queueSize: 5000
 
+    # When enabled, it uses file_storage extension to persist the queue data.
+    # This can be used to prevent data-loss during restart
```
Comment (jvoravong) [inline helm-charts/splunk-otel-collector/values.yaml:110]:
```suggestion
    # This can be used as an alternative redundancy mechanism for data being exported.
```

File: helm-charts/splunk-otel-collector/values.yaml (hunk 5 above)
Comment (wojtekzyla) [reply helm-charts/splunk-otel-collector/values.yaml:114]: As I checked the values.yaml file, `persistentQueueEnabled` is already moved under `sendingQueue `.

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 3 above)
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:650]: Why don't keep this inside the `splunk-otel-collector.splunkPlatformLogsExporter` template? This became a bit hard to read

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 7]
```diff
@@ -639,9 +644,17 @@ exporters:
...
   {{- end }}
   {{- if (eq (include "splunk-otel-collector.platformMetricsEnabled" .) "true") }}
   {{- include "splunk-otel-collector.splunkPlatformMetricsExporter" . | nindent 2 }}
+  {{- include "splunk-otel-collector.splunkPlatformSendingQueue" . | nindent 4 }}
+      {{- if .Values.splunkPlatform.sendingQueue.persistentQueueEnabled.metrics }}
+      storage: file_storage/persistent_queue
+      {{- end }}
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:657]: Same here

File: helm-charts/splunk-otel-collector/values.yaml (hunk 5 above)
Comment (jvoravong) [reply helm-charts/splunk-otel-collector/values.yaml:114]: My mistake.

File: helm-charts/splunk-otel-collector/templates/daemonset.yaml (hunk 4 above)
Comment (jvoravong) [reply helm-charts/splunk-otel-collector/templates/daemonset.yaml:163]: I think this can be confusing for users without more documentation or using different values to enable this feature.
Persistent queues can be used in the logs, metrics, and traces pipelines of the agent, cluster receiver and gateway. We need to clearly show what is supported in the values.yaml file or in an md file like [advanced-configuration.md ](https://github.com/signalfx/splunk-otel-collector-chart/blob/main/docs/advanced-configuration.md).

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 3 above)
Comment (wojtekzyla) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:650]: This is because `splunk-otel-collector.splunkPlatformLogsExporter` is used in all three agent, cluster receiver and gateway. The problem is that currently we decided to add persistent queue only in agent, because using it in others caused some errors, that no one I contacted was able to resolve.

File: helm-charts/splunk-otel-collector/templates/daemonset.yaml (hunk 4 above)
Comment (wojtekzyla) [reply helm-charts/splunk-otel-collector/templates/daemonset.yaml:163]: I added description what is supported in values.yaml [here](https://github.com/signalfx/splunk-otel-collector-chart/blob/bb3588c3843c12c253410567063c8609f133bb57/helm-charts/splunk-otel-collector/values.yaml#L113)

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 3 above)
Comment (dmitryax) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:650]: > The problem is that currently we decided to add persistent queue only in agent, because using it in others caused some errors, that no one I contacted was able to resolve.

I don't think it's a valid reason to introduce the inconsistency

Comment (wojtekzyla) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:650]: I created a new issue now #800

**Comment by VihasMakwana** [comment]
Issue https://github.com/signalfx/splunk-otel-collector-chart/issues/800 has been fixed and cluster receiver with persistent buffering is working.
